### PR TITLE
feat: brownie project support and dependencies that have dependencies

### PR DESCRIPTION
**Pull request by unparalleled-js**
### What I did

fixes: #640 

### How I did it

* refactor: split up `managers/project` to a directory space with `types` containing the `ProjectAPI` implementations, `dependency` containing the `DependencyManager`, `GithubDependency`, and `LocalDependency` and `manager` containing the `ProjectManager`
* feat: Create new class `BrownieProject(ProjectAPI)` that migrates `brownie-config.yaml` to an Ape config and creates the manifest from there exactly the same
* fix: make sure to download dependencies of a dependency
* test: test that you can support dependencies that have dependencies

### How to verify it

I used this config which I believe is the same as the one from the issue:
```yaml
plugins:
  - name: solidity
  - name: vyper
  - name: hardhat

# require OpenZepplin Contracts
dependencies:
  - name: openzeppelin
    github: OpenZeppelin/openzeppelin-contracts
    version: 4.5.0
  - name: vault
    github: yearn/yearn-vaults
    version: 0.4.3

solidity:
  import_remapping:
    - "@openzeppelin/contracts=openzeppelin/v4.5.0"

ethereum:
  local:
    default_provider: hardhat
```

I created a contract that uses the OZ dependency but not the vaults, as that seems like the use-case in the ticket:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.2;

import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
import "@openzeppelin/contracts/token/ERC721/extensions/ERC721Enumerable.sol";
import "@openzeppelin/contracts/token/ERC721/extensions/ERC721URIStorage.sol";
import "@openzeppelin/contracts/access/Ownable.sol";
import "@openzeppelin/contracts/utils/Counters.sol";

contract PoofPoof is ERC721, ERC721URIStorage, Ownable {
    using Counters for Counters.Counter;

    Counters.Counter private _tokenIdCounter;
    string public poofBaseURI;

    constructor(string memory baseURI) ERC721("PoofPoof", "POOFPOOF") {
        poofBaseURI = baseURI;
    }

    function _baseURI() internal view override returns (string memory) {
        return poofBaseURI;
    }

    function safeMint(address to, string memory uri) public onlyOwner {
        uint256 tokenId = _tokenIdCounter.current();
        _tokenIdCounter.increment();
        _safeMint(to, tokenId);
        _setTokenURI(tokenId, uri);
    }

    // The following functions are overrides required by Solidity.

    function _burn(uint256 tokenId) internal override(ERC721, ERC721URIStorage) {
        super._burn(tokenId);
    }

    function tokenURI(uint256 tokenId)
        public
        view
        override(ERC721, ERC721URIStorage)
        returns (string memory)
    {
        return super.tokenURI(tokenId);
    }
}

```

Notice that I am using a different version of OZ than `yearn-vaults`. This should work still!

(NOTE: Before I compile, I made sure I was in a clean instance. This shouldn't matter but I just like to do this anyway:

```bash
rm -rf ~/.ape/packages
rm -rf contracts/.cache
rm -rf .build/
# Could replace last two lines with `-f` in the compile command
```

Now, compile:

```bash
ape compile
```

It first downloads the dependencies, including the dependency of the dependency, which is a different version of OZ than the base-project.

the command should be successful.

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: src/ape/managers/project/dependency.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            if key in config_dependency_data:
+                return dependency_cls(
+                    **config_dependency_data,
+                )  # type: ignore
+
+        dep_id = config_dependency_data.get("name", json.dumps(config_dependency_data))
+        raise ProjectError(f"No installed dependency API that supports '{dep_id}'.")
+
+
+class LocalDependency(DependencyAPI):
```
Comment (fubuloubu) [inline src/ape/managers/project/dependency.py:100]: Why is this below `DependencyManager`, but `GithubDependency` above it?

**Review by fubuloubu — APPROVED** [review]

**Comment by unparalleled-js** [comment]
Question: Is there a better way to organize the project namespace?

**Review by fubuloubu — APPROVED** [review]

**Comment by fubuloubu** [comment]
> Question: Is there a better way to organize the project namespace?

Seems acceptable. Is there something you're unhappy with?

**Comment by unparalleled-js** [comment]
> Seems acceptable. Is there something you're unhappy with?

I bounced back and forth between the following structures:

(Option A: The way it is now)
* `managers/projects/manager` that contains `ProjectManager`
* `managers/projects/dependency` that contains `DependencyManager`, `GithubDependency`, and `LocalDependency`
* `managers/projects/types` that contains `ApeProject` and `BrownieProject`

(Option B)
* `managers/projects/managers` that contain both `ProjectManager` and `DependencyManager`
* `managers/projects/types` that contains  `GithubDependency`, `LocalDependency`, `ApeProject` and `BrownieProject`

(Option C):
* `managers/project` that contains `ProjectManager`, `ApeProject`, and `BrownieProject`
* `managers/dependency` that contains `DependencyManager`, `LocalDependency` and `GithubDependency`

**Comment by fubuloubu** [comment]
I think it's fine as is

**Pull request merged**